src/library/cratefeature.cpp

```diff
@@ -128,17 +143,71 @@ int CrateFeature::crateIdFromIndex(QModelIndex index) {
 
     QString dataPath = item->dataPath().toString();
     bool ok = false;
-    int playlistId = dataPath.toInt(&ok);
+    int crateId = dataPath.toInt(&ok);
     if (!ok) {
         return -1;
     }
-    return playlistId;
+    return crateId;
+}
+
+QModelIndex CrateFeature::indexFromCrateId(int id) {
+
+    QStack<QModelIndex> cratesToProcess;
+    QModelIndex crate;
+    cratesToProcess.push(QModelIndex());
+    while (! cratesToProcess.empty()) {
+        crate = cratesToProcess.pop();
+
+        if (id == crateIdFromIndex(crate)) {
+            qDebug() << "Found index of crate found id: " << id;
+            return crate;
+        }
+
+        for(int row = 0; row < m_childModel.rowCount(crate); row++) {
+            cratesToProcess.push(m_childModel.index(row, 0, crate));
+        }
+    }
+    qDebug() << "Failed search for index of crate id: " << id;
+    return QModelIndex();
+}
+
+// updates hasChildren function in childModel
+void CrateFeature::refreshModelHasChildren() { 
+    //put the set of folders to variable to be lambda-captureable
+    QSet<int> folderSet = m_crateDao.getFolderSet();
+    m_childModel.setHasChildrenFunction( 
+        [folderSet] (QVariant crateId) {
+             return folderSet.contains(crateId.toInt());
+        }
```

Comment:
It is not straigh way how to do it. I saw generally two other ways of doing the trick:
1) construct full childmodel at once, no lazy loading -> no need to implement hasChildren function
2) derive treeItemModel to accept list of folders and hard code hasChildren function inside

The solution I used allows any outer object to set how the presence of children is resolved without need to derive treeitemmodel for different usage.

It you think it is better to use other options I will do it that way